### PR TITLE
Chisel Linting Framework

### PR DESCRIPTION
**Pull request by azidar**
**Type of change**: new feature

**Impact**: API Addition

**Development Phase**: implementation

**Release Notes**

 - Added CHISEL_OPTIONS environment variable
 - Added FIRRTL_OPTIONS environment variable
 - Added Chisel linting options (--lint, --lint-options, --lint-whitelist:anon-regs, --lint-whitelist:trunc-widths)

**Summary**

There have been many instances where consumers of rocket-chip Verilog have to run LEC tools, debuggers, physical design tools. When registers don't have stable names, changes to the generator cause changes to the names of registers, making the lives of those downstream more difficult.

This PR adds the following:

 - A linting check which looks for unnamed registers and reports them to the user as an error
 - A linting check for connections from a larger-width signal to a smaller-width signal

To try it out:

```
> cd emulator
> CHISEL_OPTIONS="--lint * --lint-options display:*=5" make
...
Exception in thread "main" freechips.rocketchip.linting.LintException: 
Lint rule anon-regs: 102 exceptions!
 - Recommended fix:
     Use named intermediate val, or if that fails use @chiselName or *.suggestName(...)
 - Whitelist file via Chisel cmdline arg:
     --lint-whitelist:anon-regs Arbiter.scala,AtomicAutomata.scala,BTB.scala,Broadcast.scala,CSR.scala,Counter.scala,Counters.scala,DCache.scala,Debug.scala,Deinterleaver.scala,Edges.scala,FIFOFixer.scala,FPU.scala,Fragmenter.scala,Frontend.scala,HellaCacheArbiter.scala,ICache.scala,Interrupts.scala,Monitor.scala,Periphery.scala,Plic.scala,Reg.scala,Replacement.scala,RocketCore.scala,RocketTile.scala,ShiftQueue.scala,SourceShrinker.scala,ToAXI4.scala,ToTL.scala,Valid.scala,Xbar.scala,package.scala
 - Whitelist file via Chisel scala API:
     whitelist("anon-regs", Set("Arbiter.scala","AtomicAutomata.scala","BTB.scala","Broadcast.scala","CSR.scala","Counter.scala","Counters.scala","DCache.scala","Debug.scala","Deinterleaver.scala","Edges.scala","FIFOFixer.scala","FPU.scala","Fragmenter.scala","Frontend.scala","HellaCacheArbiter.scala","ICache.scala","Interrupts.scala","Monitor.scala","Periphery.scala","Plic.scala","Reg.scala","Replacement.scala","RocketCore.scala","RocketTile.scala","ShiftQueue.scala","SourceShrinker.scala","ToAXI4.scala","ToTL.scala","Valid.scala","Xbar.scala","package.scala"))
 - Disable this linting check:
     Omit anon-regs from --lint option
 - Modify display settings with:
     --lint-options ...,display:anon-regs=<number>,...
anon-regs.1: @[Arbiter.scala 21:23]  in Set(TLXbar_7, AXI4ToTL, TLXbar_5, TLXbar, TLXbar_8)
anon-regs.2: @[Arbiter.scala 54:30]  in Set(TLXbar_7, AXI4ToTL, TLBroadcast, TLError_1, TLXbar_5, TLXbar, TLXbar_8, TLAtomicAutomata_1)
anon-regs.3: @[Arbiter.scala 79:26]  in Set(TLXbar_7, AXI4ToTL, TLBroadcast, TLError_1, TLXbar_5, TLXbar, TLXbar_8, TLAtomicAutomata_1)
anon-regs.4: @[AtomicAutomata.scala 76:28]  in Set(TLAtomicAutomata_1)
anon-regs.5: @[AtomicAutomata.scala 77:24]  in Set(TLAtomicAutomata_1)

Lint rule trunc-widths: 308 exceptions!
 - Recommended fix:
     Truncate width prior to connections
 - Whitelist file via Chisel cmdline arg:
     --lint-whitelist:trunc-widths BTB.scala,Broadcast.scala,BundleMap.scala,CSR.scala,Counters.scala,Custom.scala,DCache.scala,Debug.scala,Deinterleaver.scala,DivSqrtRecFN_small.scala,Edges.scala,FIFOFixer.scala,FPU.scala,Fragmenter.scala,Frontend.scala,HellaCache.scala,HellaCacheArbiter.scala,IBuf.scala,Monitor.scala,MulAddRecFN.scala,PMP.scala,PTW.scala,Plic.scala,PlusArg.scala,RegisterRouter.scala,RocketCore.scala,SimAXIMem.scala,SourceShrinker.scala,TLB.scala,ToAXI4.scala
 - Whitelist file via Chisel scala API:
     whitelist("trunc-widths", Set("BTB.scala","Broadcast.scala","BundleMap.scala","CSR.scala","Counters.scala","Custom.scala","DCache.scala","Debug.scala","Deinterleaver.scala","DivSqrtRecFN_small.scala","Edges.scala","FIFOFixer.scala","FPU.scala","Fragmenter.scala","Frontend.scala","HellaCache.scala","HellaCacheArbiter.scala","IBuf.scala","Monitor.scala","MulAddRecFN.scala","PMP.scala","PTW.scala","Plic.scala","PlusArg.scala","RegisterRouter.scala","RocketCore.scala","SimAXIMem.scala","SourceShrinker.scala","TLB.scala","ToAXI4.scala"))
 - Disable this linting check:
     Omit trunc-widths from --lint option
 - Modify display settings with:
     --lint-options ...,display:trunc-widths=<number>,...
trunc-widths.1: @[:freechips.rocketchip.system.DefaultConfig.fir@208657.8] _T_3704 <= _T_3702 // Connecting width 64 to width 16 in Set(CSRFile)
trunc-widths.2: @[:freechips.rocketchip.system.DefaultConfig.fir@208749.8] _T_3733 <= wdata // Connecting width 64 to width 32 in Set(CSRFile)
trunc-widths.3: @[:freechips.rocketchip.system.DefaultConfig.fir@208883.8] _T_3794 <= _T_3792 // Connecting width 64 to width 16 in Set(CSRFile)
trunc-widths.4: @[:freechips.rocketchip.system.DefaultConfig.fir@209238.8] _T_3978 <= _T_3976 // Connecting width 64 to width 8 in Set(CSRFile)
trunc-widths.5: @[:freechips.rocketchip.system.DefaultConfig.fir@209274.8] _T_3998 <= _T_3996 // Connecting width 56 to width 8 in Set(CSRFile)
Makefrag-verilator:16: recipe for target '/scratch/adami/work/rocket-chip-dev/rocket-chip-linting/emulator/generated-src/freechips.rocketchip.system.DefaultConfig.v' failed
```

To see the compiler options, run with:

```
> cd emulator
> CHISEL_OPTIONS="--help" make
...
freechips.rocketchip.linting.LintReporter
  --lint [*]|[<lintRule>,<lintRule>,...]
                           Enable linting for specified rules, where * is all rules. Available rules: anon-regs,trunc-widths.
  --lint-options (strict|warn)[,displayTotal=<numError>][,display:<lintName>=<numError>]
                           Customize linting options, including strict/warn or number of violations displayed.
freechips.rocketchip.linting.rule.LintAnonymousRegisters
  --lint-whitelist:anon-regs <filename1>.scala[,<filename2>.scala]*
                           Enable linting anonymous registers for all files except provided files.
freechips.rocketchip.linting.rule.LintTruncatingWidths
  --lint-whitelist:trunc-widths <filename1>.scala[,<filename2>.scala]*
                           Enable linting anonymous registers for all files except provided files.`1
...
```

TODO:

 - [x] Add scaladoc to all files
 - [x] Solicit feedback from the community on ideas/implementation

### CODE REVIEW

**Comment by jackkoenig** [comment]
This looks great overall. I would be more specific than just "linting" since there are other possible lint rules we may want to add some day. I would call this `--lint-register-names` or something like that.

**Comment by aswaterman** [comment]
Can't we fix 80+% of these with `@chiselName` instead of the more verbose `suggestName` approach?

**Comment by aswaterman** [comment]
... or perhaps the `suggestName` ones are the ones where `@chiselName` doesn't cut it.

**Comment by jackkoenig** [comment]
> ... or perhaps the suggestName ones are the ones where @chiselName doesn't cut it.

Yeah I think `@chiselName` is missing them, bug to-be-fixed but no reason to suffer with bad register names in the meantime.

**Comment by hcook** [comment]
Seems like an ok way to document `@chiselName` failures; @azidar can confirm but I think he attempted `@chiselName` first in most cases.

**Comment by azidar** [comment]
Yeah, @chiselName didn't fix the majority of these. I think there is a serious bug, but in the meantime this PR highlights where we can fix these.

File: src/main/scala/tilelink/Monitor.scala [hunk 1]
```diff
@@ -527,14 +527,16 @@ class TLMonitor(args: TLMonitorArgs, monitorDir: MonitorDirection = MonitorDirec
     }
   }
 
+  @chiselName
```
Comment (mwachs5) [inline src/main/scala/tilelink/Monitor.scala:530]: Did this do anything? Seeing as you  still had to add all the suggestName below?

File: src/main/scala/util/Counters.scala [hunk 2]
```diff
@@ -44,11 +44,13 @@ case class WideCounter(width: Int, inc: UInt = UInt(1), reset: Boolean = true)
   private val isWide = width > 2*inc.getWidth
   private val smallWidth = if (isWide) inc.getWidth max log2Up(width) else width
   private val small = if (reset) Reg(init=UInt(0, smallWidth)) else Reg(UInt(width = smallWidth))
+  small.suggestName("small")
   private val nextSmall = small +& inc
   small := nextSmall
 
   private val large = if (isWide) {
     val r = if (reset) Reg(init=UInt(0, width - smallWidth)) else Reg(UInt(width = width - smallWidth))
+    r.suggestName("r")
```
Comment (mwachs5) [inline src/main/scala/util/Counters.scala:53]: I see your naming algorithm , but maybe “small_reg” is a better name...

File: src/main/scala/linting/ChiselLinting.scala [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+           .map { case (lint: LintError, idx: Int) => s"$idx. ${lint.info} in ${lint.modules}"}
+           .mkString("\n")}
+     """.stripMargin
+)
+
+case class LintWhitelist(whiteList: Set[String]) extends NoTargetAnnotation with ChiselAnnotation {
+  override def toFirrtl = this
+}
+
+final class ChiselLinting extends Transform with RegisteredLibrary {
```
Comment (jackkoenig) [inline src/main/scala/linting/ChiselLinting.scala:33]: Moving my feedback from generic comment to here so it can be discussed in one spot:

I would be more specific than just "linting" since there are other possible lint rules we may want to add some day. I would call this `RegNameLinting` or something like that.

File: src/main/scala/rocket/HellaCacheArbiter.scala [hunk 4]
```diff
@@ -6,6 +6,7 @@ package freechips.rocketchip.rocket
 import Chisel._
 import freechips.rocketchip.config.Parameters
 
+
```
Comment (mwachs5) [inline src/main/scala/rocket/HellaCacheArbiter.scala:9]:
```suggestion
```

unnecessary whitespace change

File: src/main/scala/rocket/RocketCore.scala [hunk 5]
```diff
@@ -882,9 +882,9 @@ class Rocket(tile: RocketTile)(implicit p: Parameters) extends CoreModule()(p)
   coreMonitorBundle.wrdst := wb_waddr
   coreMonitorBundle.wrdata := rf_wdata
   coreMonitorBundle.rd0src := wb_reg_inst(19,15)
-  coreMonitorBundle.rd0val := Reg(next=Reg(next=ex_rs(0)))
+  coreMonitorBundle.rd0val := Reg(next=Reg(next=ex_rs(0)).suggestName("ex_rs_0_reg")).suggestName("ex_rs_0_reg_reg")
```
[the remark this answers is not in the document]
Comment (mwachs5) [reply src/main/scala/rocket/RocketCore.scala:885]: ah, I was particularly thinking about `ShiftRegInit` which I added here: https://github.com/chipsalliance/rocket-chip/blob/master/src/main/scala/util/ShiftReg.scala#L10

We could think about making the Chisel ShiftReg take an optional name argument.

File: src/main/scala/linting/ChiselLinting.scala (hunk 3 above)
Comment (azidar) [reply src/main/scala/linting/ChiselLinting.scala:33]: So I imagine we could have one ChiselLinting transform, which does multiple linting passes internally, no?

**Comment by azidar** [comment]
Ok so my plan is to make this PR only add the Linting transform and options, but not run the linter by default and not include any of the name fixes so all the regressions pass. That will let us incrementally change names but still get this PR in.

File: src/main/scala/linting/LintAnonymousRegisters.scala [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+     |Please fix using @chiselName, *.suggestName(...). Otherwise, whitelist the file via the Chisel commandline option:
+     |    --linting-whitelist <filename>.scala
+     |or via annotations in your Chisel code:
+     |    annotate(LintAnonymousRegsWhitelist("<filename>.scala")).
+     |
+     |Only showing first few exceptions:
+     |${seq.zip(0 until 20)
+           .map { case (lint: LintError, idx: Int) => s"$idx. ${lint.info} in ${lint.modules}"}
+           .mkString("\n")}
+     """.stripMargin
```
Comment (jackkoenig) [inline src/main/scala/linting/LintAnonymousRegisters.scala:26]: Why only show the first 20? If we trim by default we should provide an option to see them all.

File: src/main/scala/linting/LintAnonymousRegisters.scala [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+      helpText = "Enable linting anonymous registers for all files except provided files.",
+      helpValueName = Some("<filename1>.scala[,<filename2>.scala]*")
+    )
+  )
+
+  override def inputForm = HighForm
+
+  override def outputForm = HighForm
+
+  type Errors = mutable.HashMap[FileInfo, LintError]
```
Comment (jackkoenig) [inline src/main/scala/linting/LintAnonymousRegisters.scala:58]: I would make this private

File: src/main/scala/linting/LintAnonymousRegisters.scala [hunk 8]
```diff
@@ -0,0 +1,134 @@
...
+import firrtl.annotations.NoTargetAnnotation
+import firrtl.options.{OptionsException, RegisteredLibrary, ShellOption}
+import firrtl.stage.RunFirrtlTransformAnnotation
+import chisel3.experimental.ChiselAnnotation
+
+
+import scala.collection.mutable
+
+case class LintError(info: FileInfo, modules: Set[String])
+case class LintExceptions(seq: Seq[LintError]) extends FirrtlUserException(
```
Comment (jackkoenig) [inline src/main/scala/linting/LintAnonymousRegisters.scala:15]: Thanks for making the other names more specific, but these two are still pretty generic, what is the plan for adding other lint errors?

File: src/main/scala/linting/LintAnonymousRegisters.scala [hunk 9]
```diff
@@ -0,0 +1,134 @@
...
+  s"""Linting Report: ${seq.size} exceptions!
+     |Please fix using @chiselName, *.suggestName(...). Otherwise, whitelist the file via the Chisel commandline option:
```
Comment (mwachs5) [inline src/main/scala/linting/LintAnonymousRegisters.scala:17]: i  think you can also make a suggestion to assign to an intermediate named `val` as that is the most natural thing

File: src/main/scala/linting/LintAnonymousRegisters.scala (hunk 6 above)
Comment (mwachs5) [reply src/main/scala/linting/LintAnonymousRegisters.scala:26]: if we only show the first 20 or either way, can you make it obvious how to see the reset? Something like Only showing the first few execptions, see ${filepath} for full report"?

File: src/main/scala/diplomaticobjectmodel/logicaltree/RocketLogicalTreeNode.scala [hunk 10]
```diff
@@ -83,7 +84,7 @@ class RocketLogicalTreeNode(
     val omBusError = components.collectFirst { case x: OMBusError => x }
 
     Seq(OMRocketCore(
-      isa = OMISA.rocketISA(tile, XLen),
+      isa = OMISA.rocketISA(tile, XLen, PgLevels),
```
Comment (mwachs5) [inline src/main/scala/diplomaticobjectmodel/logicaltree/RocketLogicalTreeNode.scala:87]: why is this in this PR? Maybe we should be making this against the 3.3 bump to see more clearly what is added by the linter and what is added by the chisel bump?

File: src/main/scala/diplomaticobjectmodel/model/OMISA.scala [hunk 11]
```diff
@@ -82,10 +82,11 @@ object OMISA {
     }
 
     val addressTranslationModes = xLen match {
-        case _ if !coreParams.useVM => Bare
-        case 32 => Sv32
-        case 64 => Sv39
-        case _ => throw new IllegalArgumentException(s"ERROR: Invalid Xlen: $xLen")
+      case _ if !coreParams.useVM => Bare
```
Comment (mwachs5) [inline src/main/scala/diplomaticobjectmodel/model/OMISA.scala:85]: confused why this change is in this PR

Comment (azidar) [reply src/main/scala/diplomaticobjectmodel/model/OMISA.scala:85]: yeah its because I needed 3.3 stuff. I'll make the PR against chisel-3-3

File: src/main/scala/linting/LintAnnotation.scala [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+import firrtl.ir.{Info, FileInfo}
+import firrtl.annotations.NoTargetAnnotation
+import chisel3.experimental.ChiselAnnotation
+
+/** Parent trait for all linting annotations */
+trait LintAnnotation extends NoTargetAnnotation with ChiselAnnotation {
+  override def toFirrtl = this
+}
+
+/** Represents a linting violation underr a given linter rule */
```
Comment (azidar) [inline src/main/scala/linting/LintAnnotation.scala:14]:
```suggestion
/** Represents a linting violation under a given linter rule */
```

File: src/main/scala/linting/LintAnnotation.scala [hunk 13]
```diff
@@ -0,0 +1,36 @@
...
+      case f: FileInfo => f.info.serialize match {
+        case scala(file) => Some(file)
+        case other => None
+      }
+      case other => None
+    }
+  }
+}
+
+/** A list of files to ignore lint violations on, fora  given lint rule */
```
Comment (azidar) [inline src/main/scala/linting/LintAnnotation.scala:28]:
```suggestion
/** A list of files to ignore lint violations on, for a given lint rule */
```

**Comment by jackkoenig** [comment]
@azidar hey sorry I force pushed the `chisel-3-3` branch so you'll need to rebase, I didn't realize this was against that branch. Why is this PRed against that branch instead of master? Does it rely on Chisel 3.3 stuff?
